my_compassion/controllers/my2_donations.py

```diff
@@ -0,0 +1,269 @@
+
+from werkzeug.exceptions import BadRequest, NotFound
+
+from odoo import http
+from odoo.http import request
+
+
+class MyCompassionDonationsController(http.Controller):
+    @http.route(
+        '/my2/gifts/<model("product.template"):product>',
+        type="http",
+        auth="user",
+        website=True,
+        sitemap=False,
+    )
+    def my2_render_donation_details_page(self, product, **kwargs):
+        """
+        Renders a donation details page for a specific donation (product).
+        return: An HTTP response containing a rendered template with the donation details page.
+        """
+        sponsorships = request.env.user.partner_id.sponsorship_ids
+
+        return request.render(
+            "my_compassion.my2_donation_details_page",
+            {
+                "product": product,
+                "sponsorships": sponsorships,
+            },
+        )
+
+    @http.route(
+        "/my2/gifts/new",
+        type="json",
+        auth="user",
+        website=True,
+        methods=["POST"],
+    )
+    def donation_new(self, **post):
+        """
+        Receives the donation submission and add it to the gift package of the user,
+        then redirects to checkout (the gift package page).
+        """
+        # Fetch the product record from the database
+        try:
+            product_template_id = int(post.get("product_id"))
+        except (ValueError, TypeError):
+            raise BadRequest()
+        product_template = (
+            request.env["product.template"].sudo().browse(product_template_id)
+        )
+
+        # Make sure the product is available
+        if not product_template.activate_for_my_compassion:
+            raise NotFound()
+
+        # Get current cart content
+        order = request.website.sale_get_order(force_create=True)
+
+        # Add product to the cart
+        order.write(
+            {
+                "order_line": [
+                    (
+                        0,
+                        0,
+                        self._extract_donation_order_line_fields(
+                            product_template, post
+                        ),
+                    )
+                ]
+            }
+        )
+
+    @http.route(
+        "/my2/gifts/edit",
+        type="json",
+        auth="user",
+        website=True,
+        methods=["POST"],
+    )
+    def donation_edit(self, **post):
+        """
+        Edits a donation from the user's gift package.
+        """
+        # Fetch the order line record from the database
+        order_line_id = post.get("order_line_id")
+        order_line = request.env["sale.order.line"].sudo().browse(order_line_id)
+
+        # Make sure the order line exists
+        if not order_line:
+            raise NotFound()
+
+        product_template = order_line.product_template_id
+
+        order_line.write(
+            self._extract_donation_order_line_fields(product_template, post)
+        )
```

Comment:
I would suggest to retrieve the `order_line` record from the user session. This ensure that the order being edited belongs to the current user's cart session and removes one call to the database.

Here is an example of my suggestion:
```suggestion
    def donation_edit(self, **post):
        """
        Edits a donation from the user's gift package.
        """
        # Retrieves current user's cart
        order = request.website.sale_get_order()
        
        # Get the order to be edited
        try:
            order_line_id = int(post.get("order_line_id"))
            order_line = order.order_line.filtered(lambda line: line.id == order_line_id)
        except (ValueError, TypeError):
            raise BadRequest()
            
        product_template = order_line.product_template_id

        order_line.write(
            self._extract_donation_order_line_fields(product_template, post)
        )
```

Reply:
That's indeed better. However I will leave the check for whether the order line exists.